File: packages/mobile_agent_auth/ios/mobile_agent_auth/Sources/mobile_agent_auth/NativeLlmTransportController.swift
import AppAuth
import Flutter
import Foundation

final class NativeLlmTransportController: NSObject, FlutterStreamHandler {
  private let stateStore: KeychainAuthStateStore
  private let lock = NSLock()
  private let authStateLock = NSLock()
  private let onReauthenticationRequired: () -> Void
  private var requests: [String: ActiveNativeRequest] = [:]
  private var eventSink: FlutterEventSink?
  private var cachedAuthState: OIDAuthState?

  init(
    stateStore: KeychainAuthStateStore,
    onReauthenticationRequired: @escaping () -> Void
  ) {
    self.stateStore = stateStore
    self.onReauthenticationRequired = onReauthenticationRequired
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "startStream":
      startStream(arguments: call.arguments as? [String: Any], result: result)
    case "cancelRequest":
      cancelRequest(arguments: call.arguments as? [String: Any], result: result)
    case "requestState":
      requestState(arguments: call.arguments as? [String: Any], result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func startStream(arguments: [String: Any]?, result: @escaping FlutterResult) {
    guard currentEventSink() != nil else {
      result(error("event_listener_unavailable", "Event listener is not active."))
      return
    }
    let input: NativeStreamInput
    do {
      input = try NativeStreamInput(arguments: arguments)
    } catch _ {
      result(error("request_invalid", "Streaming request is invalid."))
      return
    }
    let active = ActiveNativeRequest()
    lock.lock()
    let inserted = requests[input.requestId] == nil
    if inserted { requests[input.requestId] = active }
    lock.unlock()
    guard inserted else {
      result(error("request_in_progress", "Request is already active."))
      return
    }
    let state: OIDAuthState
    do {
      guard let restored = try currentAuthState(), restored.isAuthorized else {
        remove(requestId: input.requestId, expected: active)
        onReauthenticationRequired()
        result(error("reauthentication_required", "A valid login is required."))
        return
      }
      state = restored
    } catch _ {
      remove(requestId: input.requestId, expected: active)
      result(error("storage_failure", "Secure storage failed."))
      return
    }

    state.performAction { [weak self] accessToken, _, authError in
      guard let self else { return }
      guard !active.isCancelled else {
        self.remove(requestId: input.requestId, expected: active)
        result(self.error("request_cancelled", "Request was cancelled."))
        return
      }
      guard authError == nil, let accessToken, !accessToken.isEmpty else {
        self.remove(requestId: input.requestId, expected: active)
        self.onReauthenticationRequired()
        result(self.error("reauthentication_required", "Token refresh failed."))
        return
      }
      do {
        try self.stateStore.write(state)
      } catch {
        self.remove(requestId: input.requestId, expected: active)
        result(self.error("storage_failure", "Secure storage failed."))
        return
      }
      let task = NativeStreamTask(
        input: input,
        accessToken: accessToken,
        onEvent: { [weak self] type, data in
          if type == "error", data["code"] as? String == "reauthentication_required" {
            self?.onReauthenticationRequired()
          }
          self?.emit(requestId: input.requestId, type: type, data: data)
        },
        onComplete: { [weak self] in
          self?.remove(requestId: input.requestId, expected: active)
        }
      )
      guard active.install(task: task) else {
        self.remove(requestId: input.requestId, expected: active)
        task.cancel(sendServerCancel: false)
        result(self.error("request_cancelled", "Request was cancelled."))
        return
      }
      task.start()
      result(nil)
    }
  }

  private func cancelRequest(arguments: [String: Any]?, result: @escaping FlutterResult) {
    guard
      let requestId = arguments?["requestId"] as? String,
      requestId.range(of: "^[A-Za-z0-9_-]{8,80}$", options: .regularExpression) != nil,
      let baseUrl = arguments?["bffBaseUrl"] as? String,
      (try? NativeStreamInput.validateBaseURL(baseUrl)) != nil
    else {
      result(error("request_invalid", "Request ID is invalid."))
      return
    }
    guard let active = remove(requestId: requestId) else {
      result(cancelResult(
        requestId: requestId,
        localCancelled: false,
        serverAcknowledgment: "not_required"
      ))
      return
    }
    let task = active.cancel()
    emit(requestId: requestId, type: "cancelled", data: [:])
    guard let task else {
      result(cancelResult(
        requestId: requestId,
        localCancelled: true,
        serverAcknowledgment: "not_required"
      ))
      return
    }
    task.cancel(sendServerCancel: true) { [weak self] acknowledgment in
      DispatchQueue.main.async {
        guard let self else { return }
        result(self.cancelResult(
          requestId: requestId,
          localCancelled: true,
          serverAcknowledgment: acknowledgment
        ))
      }
    }
  }

  private func requestState(arguments: [String: Any]?, result: @escaping FlutterResult) {
    guard
      let arguments,
      Set(arguments.keys).isSubset(of: Set(["requestId"])),
      let requestId = arguments["requestId"] as? String,
      requestId.range(of: "^[A-Za-z0-9_-]{8,80}$", options: .regularExpression) != nil
    else {
      result(error("request_invalid", "Request ID is invalid."))
      return
    }
    lock.lock()
    let state = requests[requestId]?.state ?? "not_found"
    lock.unlock()
    result(["requestId": requestId, "state": state])
  }

  private func cancelResult(
    requestId: String,
    localCancelled: Bool,
    serverAcknowledgment: String
  ) -> [String: Any] {
    [
      "requestId": requestId,
      "localCancelled": localCancelled,
      "serverAcknowledgment": serverAcknowledgment,
    ]
  }

  private func remove(requestId: String) -> ActiveNativeRequest? {
    lock.lock()
    defer { lock.unlock() }
    return requests.removeValue(forKey: requestId)
  }

  private func remove(requestId: String, expected: ActiveNativeRequest) {
    lock.lock()
    if requests[requestId] === expected { requests.removeValue(forKey: requestId) }
    lock.unlock()
  }

  func cancelAll() {
    lock.lock()
    let active = Array(requests.values)
    requests.removeAll()
    lock.unlock()
    active.forEach { $0.cancel()?.cancel(sendServerCancel: false) }
  }

  func updateAuthState(_ state: OIDAuthState) {
    authStateLock.lock()
    cachedAuthState = state
    authStateLock.unlock()
  }

  func clearAuthState() {
    authStateLock.lock()
    cachedAuthState = nil
    authStateLock.unlock()
  }

  func revokeSession(bffBaseURL: URL?, completion: @escaping () -> Void) {
    cancelAll()
    let state = try? currentAuthState()
    clearAuthState()
    stateStore.clear()
    guard let state else {
      DispatchQueue.main.async(execute: completion)
      return
    }

    var revocations: [URLRequest] = []
    if
      let bffBaseURL,
      let accessToken = state.lastTokenResponse?.accessToken,
      let endpoint = URL(string: "/v1/session/revoke", relativeTo: bffBaseURL)?.absoluteURL
    {
      var request = URLRequest(url: endpoint)
      request.httpMethod = "POST"
      request.timeoutInterval = 5
      request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
      revocations.append(request)
    }

    let authorizationRequest = state.lastAuthorizationResponse.request
    let configuration = authorizationRequest.configuration
    let discovery = configuration.discoveryDocument?.discoveryDictionary
    if
      let refreshToken = state.refreshToken,
      !refreshToken.isEmpty,
      let rawEndpoint = discovery?["revocation_endpoint"] as? String,
      let endpoint = URL(string: rawEndpoint),
      let issuer = configuration.issuer,
      Self.sameHostHTTPS(issuer: issuer, endpoint: endpoint)
    {
      var request = URLRequest(url: endpoint)
      request.httpMethod = "POST"
      request.timeoutInterval = 5
      request.setValue(
        "application/x-www-form-urlencoded; charset=utf-8",
        forHTTPHeaderField: "Content-Type"
      )
      request.httpBody = Self.formBody([
        "token": refreshToken,
        "token_type_hint": "refresh_token",
        "client_id": authorizationRequest.clientID,
      ])
      revocations.append(request)
    }

    guard !revocations.isEmpty else {
      DispatchQueue.main.async(execute: completion)
      return
    }
    let group = DispatchGroup()
    let session = URLSession(configuration: .ephemeral)
    for request in revocations {
      group.enter()
      session.dataTask(with: request) { _, _, _ in
        group.leave()
      }.resume()
    }
    group.notify(queue: .main) {
      session.finishTasksAndInvalidate()
      completion()
    }
  }

  private static func sameHostHTTPS(issuer: URL, endpoint: URL) -> Bool {
    issuer.scheme == "https"
      && endpoint.scheme == "https"
      && issuer.host?.caseInsensitiveCompare(endpoint.host ?? "") == .orderedSame
      && endpoint.user == nil
      && endpoint.password == nil
      && endpoint.fragment == nil
  }

  private static func formBody(_ values: [String: String]) -> Data? {
    var components = URLComponents()
    components.queryItems = values.sorted(by: { $0.key < $1.key }).map {
      URLQueryItem(name: $0.key, value: $0.value)
    }
    return components.percentEncodedQuery?.data(using: .utf8)
  }

  private func currentAuthState() throws -> OIDAuthState? {
    authStateLock.lock()
    defer { authStateLock.unlock() }
    if let cachedAuthState, cachedAuthState.isAuthorized {
      return cachedAuthState
    }
    let restored = try stateStore.read()
    cachedAuthState = restored
    return restored
  }

  private func emit(requestId: String, type: String, data: [String: Any]) {
    DispatchQueue.main.async { [weak self] in
      self?.currentEventSink()?([
        "requestId": requestId,
        "type": type,
        "data": data,
      ])
    }
  }

  private func currentEventSink() -> FlutterEventSink? {
    lock.lock()
    defer { lock.unlock() }
    return eventSink
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    lock.lock()
    eventSink = events
    lock.unlock()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    lock.lock()
    eventSink = nil
    lock.unlock()
    cancelAll()
    return nil
  }

  private func error(_ code: String, _ message: String) -> FlutterError {
    FlutterError(code: code, message: message, details: nil)
  }
}

struct NativeStreamInput {
  let requestId: String
  let baseURL: URL
  let requestData: Data

  init(arguments: [String: Any]?) throws {
    guard
      let arguments,
      Set(arguments.keys).isSubset(of: Set(["bffBaseUrl", "request"])),
      let baseUrl = arguments["bffBaseUrl"] as? String,
      let baseURL = try? Self.validateBaseURL(baseUrl),
      let request = arguments["request"] as? [String: Any],
      let requestId = request["request_id"] as? String,
      requestId.range(of: "^[A-Za-z0-9_-]{8,80}$", options: .regularExpression) != nil,
      JSONSerialization.isValidJSONObject(request)
    else { throw NativeStreamInputError.invalid }
    let requestData = try JSONSerialization.data(withJSONObject: request)
    guard requestData.count <= 500_000 else { throw NativeStreamInputError.invalid }
    self.requestId = requestId
    self.baseURL = baseURL
    self.requestData = requestData
  }

  static func validateBaseURL(_ value: String) throws -> URL {
    guard
      let components = URLComponents(string: value),
      components.scheme == "https",
      components.host != nil,
      components.user == nil,
      components.password == nil,
      components.query == nil,
      components.fragment == nil,
      let url = components.url
    else { throw NativeStreamInputError.invalid }
    return url
  }
}

private enum NativeStreamInputError: Error { case invalid }

private final class ActiveNativeRequest {
  private let lock = NSLock()
  private var task: NativeStreamTask?
  private var cancelled = false
  private var phase = "preparing"

  var isCancelled: Bool {
    lock.lock()
    defer { lock.unlock() }
    return cancelled
  }

  var state: String {
    lock.lock()
    defer { lock.unlock() }
    return phase
  }

  func install(task: NativeStreamTask) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    guard !cancelled else { return false }
    self.task = task
    phase = "streaming"
    return true
  }

  func cancel() -> NativeStreamTask? {
    lock.lock()
    defer { lock.unlock() }
    cancelled = true
    phase = "cancelling"
    return task
  }
}

private final class NativeStreamTask: NSObject, URLSessionDataDelegate {
  private let input: NativeStreamInput
  private let accessToken: String
  private let onEvent: (String, [String: Any]) -> Void
  private let onComplete: () -> Void
  private let lock = NSLock()
  private var session: URLSession?
  private var task: URLSessionDataTask?
  private var buffer = Data()
  private var eventType: String?
  private var dataLines: [String] = []
  private var terminalEmitted = false
  private var cancelled = false

  init(
    input: NativeStreamInput,
    accessToken: String,
    onEvent: @escaping (String, [String: Any]) -> Void,
    onComplete: @escaping () -> Void
  ) {
    self.input = input
    self.accessToken = accessToken
    self.onEvent = onEvent
    self.onComplete = onComplete
  }

  func start() {
    guard let url = URL(string: "/v1/chat/stream", relativeTo: input.baseURL)?.absoluteURL else {
      emitTerminal("error", ["code": "configuration_invalid"])
      onComplete()
      return
    }
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.httpBody = input.requestData
    request.timeoutInterval = 130
    request.cachePolicy = .reloadIgnoringLocalCacheData
    request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
    request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
    request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
    let queue = OperationQueue()
    queue.maxConcurrentOperationCount = 1
    let session = URLSession(configuration: .ephemeral, delegate: self, delegateQueue: queue)
    self.session = session
    let task = session.dataTask(with: request)
    self.task = task
    task.resume()
  }

  func cancel(
    sendServerCancel: Bool,
    completion: ((String) -> Void)? = nil
  ) {
    lock.lock()
    cancelled = true
    let task = self.task
    lock.unlock()
    task?.cancel()
    if sendServerCancel {
      postServerCancel(completion: completion)
    } else {
      completion?("not_required")
    }
    session?.invalidateAndCancel()
  }

  func urlSession(
    _ session: URLSession,
    dataTask: URLSessionDataTask,
    didReceive response: URLResponse,
    completionHandler: @escaping (URLSession.ResponseDisposition) -> Void
  ) {
    guard let response = response as? HTTPURLResponse else {
      emitTerminal("error", ["code": "network_unavailable"])
      completionHandler(.cancel)
      return
    }
    guard (200...299).contains(response.statusCode) else {
      emitTerminal("error", ["code": statusError(response.statusCode)])
      completionHandler(.cancel)
      return
    }
    completionHandler(.allow)
  }

  func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
    lock.lock()
    buffer.append(data)
    while let newline = buffer.firstIndex(of: 0x0A) {
      var lineData = buffer.prefix(upTo: newline)
      buffer.removeSubrange(...newline)
      if lineData.last == 0x0D { lineData = lineData.dropLast() }
      guard let line = String(data: lineData, encoding: .utf8) else {
        lock.unlock()
        emitTerminal("error", ["code": "stream_invalid"])
        task?.cancel()
        return
      }
      consume(line: line)
    }
    lock.unlock()
  }

  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didCompleteWithError error: Error?
  ) {
    lock.lock()
    let shouldEmitError = !cancelled && !terminalEmitted
    lock.unlock()
    if shouldEmitError {
      emitTerminal("error", ["code": error == nil ? "stream_ended_early" : "network_unavailable"])
    }
    session.finishTasksAndInvalidate()
    onComplete()
  }

  private func consume(line: String) {
    if line.isEmpty {
      guard !dataLines.isEmpty else { eventType = nil; return }
      let raw = dataLines.joined(separator: "\n")
      dataLines.removeAll(keepingCapacity: true)
      let type = eventType ?? "message"
      eventType = nil
      guard
        let data = raw.data(using: .utf8),
        let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
      else {
        emitTerminalLocked("error", ["code": "stream_invalid"])
        task?.cancel()
        return
      }
      if ["done", "cancelled", "error"].contains(type) {
        emitTerminalLocked(type, object)
      } else {
        onEvent(type, object)
      }
    } else if line.hasPrefix("event:") {
      eventType = String(line.dropFirst(6)).trimmingCharacters(in: .whitespaces)
    } else if line.hasPrefix("data:") {
      dataLines.append(String(line.dropFirst(5)).trimmingCharacters(in: .whitespaces))
    }
  }

  private func emitTerminal(_ type: String, _ data: [String: Any]) {
    lock.lock()
    emitTerminalLocked(type, data)
    lock.unlock()
  }

  private func emitTerminalLocked(_ type: String, _ data: [String: Any]) {
    guard !terminalEmitted else { return }
    terminalEmitted = true
    onEvent(type, data)
  }

  private func postServerCancel(completion: ((String) -> Void)?) {
    guard let url = URL(
      string: "/v1/requests/\(input.requestId)/cancel",
      relativeTo: input.baseURL
    )?.absoluteURL else {
      completion?("unavailable")
      return
    }
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.timeoutInterval = 5
    request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
    let session = URLSession(configuration: .ephemeral)
    session.dataTask(with: request) { _, response, _ in
      let status = (response as? HTTPURLResponse)?.statusCode
      let acknowledgment: String
      switch status {
      case 202: acknowledgment = "accepted"
      case 404: acknowledgment = "not_active"
      default: acknowledgment = "unavailable"
      }
      completion?(acknowledgment)
      session.finishTasksAndInvalidate()
    }.resume()
  }

  private func statusError(_ status: Int) -> String {
    switch status {
    case 401: return "reauthentication_required"
    case 403: return "provider_access_denied"
    case 404: return "provider_not_found"
    case 429: return "provider_rate_limited"
    case 500...599: return "provider_unavailable"
    default: return "request_rejected"
    }
  }
}

File: packages/mobile_agent_auth/android/src/main/kotlin/ai/coreline/mobile_agent_auth/NativeLlmTransportController.kt
package ai.coreline.mobile_agent_auth

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URI
import java.net.URLEncoder
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import javax.net.ssl.HttpsURLConnection
import net.openid.appauth.AuthorizationService
import net.openid.appauth.AuthState
import org.json.JSONArray
import org.json.JSONObject

internal class NativeLlmTransportController(
    private val stateStore: SecureAuthStateStore,
    private val authorizationService: AuthorizationService,
    private val onReauthenticationRequired: () -> Unit,
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private val executor: ExecutorService = Executors.newCachedThreadPool()
    private val requests = ConcurrentHashMap<String, ActiveRequest>()
    private val mainHandler = Handler(Looper.getMainLooper())
    private val authStateLock = Any()

    @Volatile
    private var cachedAuthState: AuthState? = null

    @Volatile
    private var eventSink: EventChannel.EventSink? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startStream" -> startStream(call.arguments as? Map<*, *>, result)
            "cancelRequest" -> cancelRequest(call.arguments as? Map<*, *>, result)
            "requestState" -> requestState(call.arguments as? Map<*, *>, result)
            else -> result.notImplemented()
        }
    }

    private fun startStream(arguments: Map<*, *>?, result: MethodChannel.Result) {
        if (eventSink == null) {
            result.error("event_listener_unavailable", "Event listener is not active.", null)
            return
        }
        val parsed = runCatching { NativeStreamInput.from(arguments) }.getOrElse {
            result.error("request_invalid", "Streaming request is invalid.", null)
            return
        }
        val active = ActiveRequest(parsed.requestId, parsed.baseUrl)
        if (requests.putIfAbsent(parsed.requestId, active) != null) {
            result.error("request_in_progress", "Request is already active.", null)
            return
        }
        val state = synchronized(authStateLock) {
            cachedAuthState?.takeIf { it.isAuthorized }
                ?: stateStore.read()?.also { cachedAuthState = it }
        }
        if (state?.isAuthorized != true) {
            requests.remove(parsed.requestId)
            onReauthenticationRequired()
            result.error("reauthentication_required", "A valid login is required.", null)
            return
        }
        state.performActionWithFreshTokens(authorizationService) { accessToken, _, error ->
            if (error != null || accessToken.isNullOrBlank() || active.cancelled.get()) {
                requests.remove(parsed.requestId)
                if (active.cancelled.get()) {
                    result.error("request_cancelled", "Request was cancelled.", null)
                } else {
                    onReauthenticationRequired()
                    result.error("reauthentication_required", "Token refresh failed.", null)
                }
                return@performActionWithFreshTokens
            }
            try {
                stateStore.write(state)
            } catch (_: SecureStoreException) {
                requests.remove(parsed.requestId)
                result.error("storage_failure", "Secure storage failed.", null)
                return@performActionWithFreshTokens
            }
            active.accessToken = accessToken
            active.phase.set("streaming")
            active.future = executor.submit { executeStream(parsed, active) }
            result.success(null)
        }
    }

    private fun cancelRequest(arguments: Map<*, *>?, result: MethodChannel.Result) {
        val requestId = arguments?.get("requestId") as? String
        val baseUrl = arguments?.get("bffBaseUrl") as? String
        if (
            requestId == null ||
            !REQUEST_ID.matches(requestId) ||
            baseUrl == null ||
            !isValidBaseUrl(baseUrl)
        ) {
            result.error("request_invalid", "Request ID is invalid.", null)
            return
        }
        val active = requests.remove(requestId)
        if (active == null) {
            result.success(cancelResult(requestId, false, "not_required"))
            return
        }
        active.phase.set("cancelling")
        active.cancelled.set(true)
        active.connection.getAndSet(null)?.disconnect()
        active.future?.cancel(true)
        emitTerminal(active, "cancelled", emptyMap())
        val token = active.accessToken
        if (token.isNullOrBlank()) {
            result.success(cancelResult(requestId, true, "not_required"))
            return
        }
        runCatching {
            executor.submit {
                val acknowledgment = sendServerCancel(baseUrl, requestId, token)
                mainHandler.post {
                    result.success(cancelResult(requestId, true, acknowledgment))
                }
            }
        }.onFailure {
            result.success(cancelResult(requestId, true, "unavailable"))
        }
    }

    private fun requestState(arguments: Map<*, *>?, result: MethodChannel.Result) {
        val requestId = arguments?.get("requestId") as? String
        if (
            arguments == null ||
            arguments.keys.any { it != "requestId" } ||
            requestId == null ||
            !REQUEST_ID.matches(requestId)
        ) {
            result.error("request_invalid", "Request ID is invalid.", null)
            return
        }
        result.success(
            mapOf(
                "requestId" to requestId,
                "state" to (requests[requestId]?.phase?.get() ?: "not_found"),
            ),
        )
    }

    private fun executeStream(input: NativeStreamInput, active: ActiveRequest) {
        var terminalReceived = false
        try {
            val connection = openConnection(
                endpoint(input.baseUrl, "/v1/chat/stream"),
                active.accessToken ?: return,
            )
            active.connection.set(connection)
            connection.outputStream.use { stream ->
                stream.write(input.requestJson.encodeToByteArray())
            }
            if (connection.responseCode !in 200..299) {
                if (connection.responseCode == 401) onReauthenticationRequired()
                emitTerminal(active, "error", mapOf("code" to statusError(connection.responseCode)))
                return
            }
            BufferedReader(InputStreamReader(connection.inputStream, Charsets.UTF_8)).use { reader ->
                var eventType: String? = null
                val data = StringBuilder()
                while (!active.cancelled.get()) {
                    val line = reader.readLine() ?: break
                    when {
                        line.isEmpty() -> {
                            if (data.isNotEmpty()) {
                                val type = eventType ?: "message"
                                val payload = jsonToMap(JSONObject(data.toString()))
                                if (type in TERMINAL_EVENTS) {
                                    terminalReceived = true
                                    emitTerminal(active, type, payload)
                                } else {
                                    emit(active.requestId, type, payload)
                                }
                                data.clear()
                                eventType = null
                            }
                        }
                        line.startsWith("event:") -> eventType = line.substring(6).trim()
                        line.startsWith("data:") -> {
                            if (data.isNotEmpty()) data.append('\n')
                            data.append(line.substring(5).trimStart())
                        }
                    }
                }
            }
            if (!active.cancelled.get() && !terminalReceived) {
                emitTerminal(active, "error", mapOf("code" to "stream_ended_early"))
            }
        } catch (_: Exception) {
            if (!active.cancelled.get()) {
                emitTerminal(active, "error", mapOf("code" to "network_unavailable"))
            }
        } finally {
            active.connection.getAndSet(null)?.disconnect()
            requests.remove(active.requestId, active)
        }
    }

    private fun sendServerCancel(
        baseUrl: String,
        requestId: String,
        accessToken: String,
    ): String = runCatching {
            val connection = openConnection(
                endpoint(baseUrl, "/v1/requests/$requestId/cancel"),
                accessToken,
            )
            connection.readTimeout = 5_000
            connection.outputStream.use { it.write(byteArrayOf()) }
            val acknowledgment = when (connection.responseCode) {
                202 -> "accepted"
                404 -> "not_active"
                else -> "unavailable"
            }
            connection.disconnect()
            acknowledgment
        }.getOrDefault("unavailable")

    private fun cancelResult(
        requestId: String,
        localCancelled: Boolean,
        serverAcknowledgment: String,
    ): Map<String, Any> = mapOf(
        "requestId" to requestId,
        "localCancelled" to localCancelled,
        "serverAcknowledgment" to serverAcknowledgment,
    )

    private fun openConnection(url: String, accessToken: String): HttpsURLConnection =
        (URI(url).toURL().openConnection() as HttpsURLConnection).apply {
            requestMethod = "POST"
            doOutput = true
            connectTimeout = 15_000
            readTimeout = 130_000
            useCaches = false
            instanceFollowRedirects = false
            setRequestProperty("Authorization", "Bearer $accessToken")
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
            setRequestProperty("Accept", "text/event-stream")
        }

    private fun endpoint(baseUrl: String, path: String): String = baseUrl.trimEnd('/') + path

    private fun emit(requestId: String, type: String, data: Map<String, Any?>) {
        val event = mapOf(
            "requestId" to requestId,
            "type" to type,
            "data" to data,
        )
        mainHandler.post { eventSink?.success(event) }
    }

    private fun emitTerminal(active: ActiveRequest, type: String, data: Map<String, Any?>) {
        if (active.terminalEmitted.compareAndSet(false, true)) {
            emit(active.requestId, type, data)
        }
    }

    override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
        eventSink = sink
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        cancelAll()
    }

    fun cancelAll() {
        requests.values.toList().forEach { active ->
            requests.remove(active.requestId, active)
            active.cancelled.set(true)
            active.phase.set("cancelling")
            active.connection.getAndSet(null)?.disconnect()
            active.future?.cancel(true)
        }
    }

    fun updateAuthState(state: AuthState) {
        synchronized(authStateLock) {
            cachedAuthState = state
        }
    }

    fun clearAuthState() {
        synchronized(authStateLock) {
            cachedAuthState = null
        }
    }

    fun revokeSession(bffBaseUrl: String?, onComplete: () -> Unit) {
        cancelAll()
        val state = synchronized(authStateLock) {
            (cachedAuthState ?: stateStore.read()).also { cachedAuthState = null }
        }
        stateStore.clear()
        if (state == null) {
            mainHandler.post(onComplete)
            return
        }
        executor.submit {
            state.accessToken?.takeIf(String::isNotBlank)?.let { accessToken ->
                bffBaseUrl?.let { revokeBffSession(it, accessToken) }
            }
            revokeIdentityProviderRefresh(state)
            mainHandler.post(onComplete)
        }
    }

    fun dispose() {
        cancelAll()
        executor.shutdownNow()
        eventSink = null
        clearAuthState()
    }

    private fun statusError(statusCode: Int): String = when (statusCode) {
        401 -> "reauthentication_required"
        403 -> "provider_access_denied"
        404 -> "provider_not_found"
        429 -> "provider_rate_limited"
        in 500..599 -> "provider_unavailable"
        else -> "request_rejected"
    }

    companion object {
        private val REQUEST_ID = Regex("^[A-Za-z0-9_-]{8,80}$")
        private val TERMINAL_EVENTS = setOf("done", "cancelled", "error")

        fun isValidBaseUrl(value: String): Boolean = runCatching {
            val uri = URI(value)
            uri.scheme == "https" &&
                !uri.host.isNullOrBlank() &&
                uri.userInfo == null &&
                uri.query == null &&
                uri.fragment == null
        }.getOrDefault(false)
    }

    private fun revokeBffSession(baseUrl: String, accessToken: String) {
        runCatching {
            val connection = openConnection(
                endpoint(baseUrl, "/v1/session/revoke"),
                accessToken,
            )
            connection.readTimeout = 5_000
            connection.outputStream.use { it.write(byteArrayOf()) }
            connection.responseCode
            connection.disconnect()
        }
    }

    private fun revokeIdentityProviderRefresh(state: AuthState) {
        val refreshToken = state.refreshToken?.takeIf(String::isNotBlank) ?: return
        val configuration = state.authorizationServiceConfiguration ?: return
        val discovery = configuration.discoveryDoc?.docJson ?: return
        val revocationEndpoint = discovery.optString("revocation_endpoint")
        val issuer = discovery.optString("issuer")
        val clientId = state.lastAuthorizationResponse?.request?.clientId ?: return
        if (!sameHostHttps(issuer, revocationEndpoint)) return
        runCatching {
            val form = formBody(
                mapOf(
                    "token" to refreshToken,
                    "token_type_hint" to "refresh_token",
                    "client_id" to clientId,
                ),
            )
            val connection = URI(revocationEndpoint).toURL().openConnection()
                as HttpsURLConnection
            connection.requestMethod = "POST"
            connection.doOutput = true
            connection.connectTimeout = 5_000
            connection.readTimeout = 5_000
            connection.useCaches = false
            connection.instanceFollowRedirects = false
            connection.setRequestProperty(
                "Content-Type",
                "application/x-www-form-urlencoded; charset=utf-8",
            )
            connection.outputStream.use { it.write(form.encodeToByteArray()) }
            connection.responseCode
            connection.disconnect()
        }
    }

    private fun sameHostHttps(issuer: String, endpoint: String): Boolean = runCatching {
        val issuerUri = URI(issuer)
        val endpointUri = URI(endpoint)
        issuerUri.scheme == "https" &&
            endpointUri.scheme == "https" &&
            !issuerUri.host.isNullOrBlank() &&
            issuerUri.host.equals(endpointUri.host, ignoreCase = true) &&
            endpointUri.userInfo == null &&
            endpointUri.fragment == null
    }.getOrDefault(false)

    private fun formBody(values: Map<String, String>): String = values.entries.joinToString("&") {
        "${URLEncoder.encode(it.key, Charsets.UTF_8.name())}=" +
            URLEncoder.encode(it.value, Charsets.UTF_8.name())
    }
}

private data class NativeStreamInput(
    val requestId: String,
    val baseUrl: String,
    val requestJson: String,
) {
    companion object {
        fun from(arguments: Map<*, *>?): NativeStreamInput {
            requireNotNull(arguments)
            require(arguments.keys.all { it in setOf("bffBaseUrl", "request") })
            val baseUrl = arguments["bffBaseUrl"] as? String ?: error("base URL missing")
            validateBaseUrl(baseUrl)
            val request = arguments["request"] as? Map<*, *> ?: error("request missing")
            val requestId = request["request_id"] as? String ?: error("request ID missing")
            require(Regex("^[A-Za-z0-9_-]{8,80}$").matches(requestId))
            val json = JSONObject(request).toString()
            require(json.length <= 500_000)
            return NativeStreamInput(requestId, baseUrl, json)
        }

        private fun validateBaseUrl(value: String) {
            require(NativeLlmTransportController.isValidBaseUrl(value))
        }
    }
}

private class ActiveRequest(
    val requestId: String,
    val baseUrl: String,
) {
    val cancelled = AtomicBoolean(false)
    val terminalEmitted = AtomicBoolean(false)
    val connection = AtomicReference<HttpsURLConnection?>()
    val phase = AtomicReference("preparing")

    @Volatile
    var accessToken: String? = null

    @Volatile
    var future: Future<*>? = null
}

private fun jsonToMap(value: JSONObject): Map<String, Any?> = value.keys().asSequence().associateWith {
    jsonValue(value.get(it))
}

private fun jsonValue(value: Any?): Any? = when (value) {
    JSONObject.NULL -> null
    is JSONObject -> jsonToMap(value)
    is JSONArray -> (0 until value.length()).map { jsonValue(value.get(it)) }
    else -> value
}
